pkg/flow/internal/controller/node_declare_component.go

```diff
@@ -0,0 +1,361 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/module"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/agent/pkg/flow/tracing"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/vm"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+)
+
+// DeclareComponentNode is a controller node which manages a module.
+//
+// DeclareComponentNode manages the underlying module and caches its current
+// arguments and exports.
+type DeclareComponentNode struct {
+	id                ComponentID
+	globalID          string
+	label             string
+	componentName     string
+	importLabel       string
+	declareLabel      string
+	nodeID            string // Cached from id.String() to avoid allocating new strings every time NodeID is called.
+	managedOpts       component.Options
+	registry          *prometheus.Registry
+	moduleController  ModuleController
+	OnComponentUpdate func(cn NodeWithDependants) // Informs controller that we need to reevaluate
+
+	GetModuleInfo  func(fullName string, importLabel string, declareLabel string) (ModuleInfo, error) // Retrieve the module config.
+	lastUpdateTime atomic.Time
+
+	mut     sync.RWMutex
+	block   *ast.BlockStmt // Current River block to derive args from
+	eval    *vm.Evaluator
+	managed *module.ModuleComponent // Inner managed module
+	args    component.Arguments     // Evaluated arguments for the managed component
+
+	// NOTE(rfratto): health and exports have their own mutex because they may be
+	// set asynchronously while mut is still being held (i.e., when calling Evaluate
+	// and the managed module immediately creates new exports)
+
+	healthMut  sync.RWMutex
+	evalHealth component.Health // Health of the last evaluate
+	runHealth  component.Health // Health of running the component
+
+	exportsMut sync.RWMutex
+	exports    component.Exports // Evaluated exports for the managed module
+}
+
+// ExtractImportAndDeclareLabels extract an importLabel and a declareLabel from a componentName.
+func ExtractImportAndDeclareLabels(componentName string) (string, string) {
+	parts := strings.Split(componentName, ".")
+	if len(parts) == 0 {
+		return "", ""
+	}
+	// If this is a local declare.
+	importLabel := ""
+	declareLabel := parts[0]
+	// If this is an imported module.
+	if len(parts) > 1 {
+		importLabel = parts[0]
+		declareLabel = parts[1]
+	}
+	return importLabel, declareLabel
+}
```

Comment:
This is making an assumption about the name of the component indicating its origin, and I don't think that'll hold true forever; it's likely that in the near future we'll be asked to support declaring components with `.` in the name so they feel even more like native components. 

Can we make this code more flexible for future-proofing or find a way to avoid it entirely?

Reply:
If we allow "." in the names, what happens when I declare a component "A.B" and I have an import statement "A" that imports a declare named "B"? Both should be instantiated via "A.B" according to the current design.

Reply:
Also I believe that rivers currently does not support "." in labels?